Calling Contracts (#1134)
FuelLabs/sway: docs/src/blockchain-development/calling_contracts.md

# Calling Contracts

Smart contracts can be _called_ by other contracts or scripts. In the FuelVM, this is done primarily with the [`call`](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#call-call-contract) instruction.

Sway provides a nice way to manage callable interfaces with its `abi` system.

## Example

> Here is an example of a *contract calling another contract* in Sway:

Reviewer: ```suggestion
Here is an example of a contract calling another contract in Sway. A script can call a contract in the same way.
```

    fn make_call();
}

const contract_id = 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0;

impl ContractB for Contract {
    fn make_call() {
      let x = abi(ContractA, contract_id);
      let return_value = x.receive(true, 3); // will be 45
    }
}
```

## Advanced Calls

All calls forward a gas stipend, and may additionally forward one native asset with the call.

Here is an example of how to specify the amount of gas (`gas`), the asset ID of the native asset (`asset_id`), and the amount of the native asset (`amount`) to forward:

```sway
script;

abi MyContract {
    fn foo(field_1: bool, field_2: u64);
}

fn main() {
    let x = abi(MyContract, 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0);
    let native_asset_id = 0x7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777;
    x.foo {
        gas: 5000, asset_id: asset_id, amount: 5000
    }
    (true, 3);
}
```

## Handling Re-entrancy

A common attack vector for smart contracts is [re-entrancy](https://docs.soliditylang.org/en/v0.8.4/security-considerations.html#re-entrancy). Similar to the Ethereum Virtual Machine, the FuelVM allows for re-entrancy.

> A _stateless_ re-entrancy guard is included in the Sway standard library.
> 
> Note, the guard will runtime panic if re-entrancy is detected.

Reviewer: ```suggestion
A _stateless_ re-entrancy guard is included in the Sway standard library. The guard will panic (revert) at run time if re-entrancy is detected.
```

```sway
contract;

use std::reentrancy::reentrancy_guard;

abi MyContract {
    fn some_method();
}

impl ContractB for Contract {
    fn some_method() {
        reentrancy_guard();
        // do something
    }
}
```

## Differences from Ethereum

> While the Fuel contract calling paradigm is similar to Ethereum's (using an ABI, forwarding gas and data), it differs in *two* key ways:

Reviewer: ```suggestion
While the Fuel contract calling paradigm is similar to Ethereum's (using an ABI, forwarding gas and data), it differs in _two_ key ways:
```

# Calling Contracts

End users and smart contracts typically need to communicate with each other.

The FuelVM achieves end user and inter contract communication primarily with the [`call`](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#call-call-contract) opcode.

> Sway provides a nice way manage calls with its `abi` system.

Reviewer 2: ```suggestion
Sway provides a nice way to manage calls with its `abi` system.
```

> End users and smart contracts typically need to communicate with each other.
> 
> The FuelVM achieves end user and inter contract communication primarily with the [`call`](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#call-call-contract) opcode.
> 
> Sway provides a nice way manage calls with its `abi` system.

Reviewer: ```suggestion
Smart contracts can be _called_ by other contracts or scripts. In the FuelVM, this is done primarily with the [`call`](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#call-call-contract) instruction.

Sway provides a nice way manage callable interfaces with its `abi` system.
```

Author: Added, but couldn't commit change due to other changes before.

## Script Calls

Here is an example of a *script calling a contract* in Sway:

> ```rs

Reviewer: use `sway` instead of `rs` for Sway snippets.

Author: Done.

script;

abi MyContract {
    fn foo(field_1: bool, field_2: u64);
}

> const contract_address = 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0;

Reviewer 2: ```suggestion
const contract_id = 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0;
```

fn main() {

>     let x = abi(MyContract, contract_address);

Reviewer 2: ```suggestion
    let x = abi(MyContract, contract_id);
```

    x.foo(true, 3);
}
```

> ## Inter Contract Calls

Reviewer: This section is basically a duplicate of the previous one. There's no real difference between calling _from_ a script or a contract, so this should simply be mentioned in a single sentence instead of duplicating an entire section.

Author: Agreed, we can merge, I just wanted to showcase that it works in both. I'll make a note.

Author: Removed.

> Here is an example of *contract calling another contract* in Sway:

Reviewer 2: ```suggestion
Here is an example of a *contract calling another contract* in Sway:
```

```rs
// ./contract_a.sw
contract;

abi ContractA {
    fn receive(field_1: bool, field_2: u64) -> u64;
}

impl ContractA for Contract {
    fn receive(field_1: bool, field_2: u64) -> u64 {
        assert(field_1 == true);
        assert(field_2 > 0);
        45
    }
}
```

```rs
// ./contract_b.sw
contract;

use contract_a::ContractA;

abi ContractB {
    fn make_call();
}

> const contract_address = 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0;

Reviewer 2: ```suggestion
const contract_id = 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0;
```

impl ContractB for Contract {
    fn make_call() {

>       let x = abi(ContractA, contract_address);

Reviewer 2: ```suggestion
      let x = abi(ContractA, contract_id);
```

      let return_value = x.receive(true, 3); // will be 45
    }
}
```

## Advanced Calls

> Calls forward gas and may push Native Assets into contracts as well.

Reviewer: ```suggestion
All calls forward a gas stipend, and may additionally forward one native asset with the call.
```

> Here is an example of how to specify the `gas`, [Native `asset_id`](./native_assets.md) and `amount` to forward:

Reviewer: ```suggestion
Here is an example of how to specify the amount of gas (`gas`), the asset ID of the native asset (`asset_id`), and the amount of the native asset (`amount`) to forward:
```

```rs
script;

abi MyContract {
    fn foo(field_1: bool, field_2: u64);
}

fn main() {
    let x = abi(MyContract, 0x79fa8779bed2f36c3581d01c79df8da45eee09fac1fd76a5a656e16326317ef0);
    let native_asset_id = 0x7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777_7777;
    x.foo {
        gas: 5000, asset_id: asset_id, amount: 5000
    }
    (true, 3);
}
```

## Handling Re-entrancy

A common attack vector for smart contracts is [re-entrancy](https://quantstamp.com/blog/what-is-a-re-entrancy-attack).

This is no different in Fuel and Sway contracts.

> Luckily for Sway developers, we provide a stateless re-entracy gaurd in our standard library:

Reviewer 2: We should probably say what the guard does. e.g. emit a warning/error.

Author: Agreed, I'll add a point there.

> A common attack vector for smart contracts is [re-entrancy](https://quantstamp.com/blog/what-is-a-re-entrancy-attack).
> 
> This is no different in Fuel and Sway contracts.
> 
> Luckily for Sway developers, we provide a stateless re-entracy gaurd in our standard library:

Reviewer: ```suggestion
A common attack vector for smart contracts is [re-entrancy](https://docs.soliditylang.org/en/v0.8.4/security-considerations.html#re-entrancy). Similar to the Ethereum Virtual Machine, the FuelVM allows for re-entrancy.

A _stateless_ re-entrancy guard in the Sway standard library.
```

```rs
contract;

> use std::reentrancy::{reentrancy_guard};

Reviewer 2: ```suggestion
use std::reentrancy::reentrancy_guard;
```

> use std::reentrancy::{reentrancy_guard};

Reviewer: ```suggestion
use std::reentrancy::reentrancy_guard;
```

abi MyContract {
    fn some_method();
}

impl ContractB for Contract {
    fn some_method() {
        reentrancy_guard();
        // do something
    }
}
```

## Differences from Ethereum

While Fuel does share some similar conceptual call paradigms to Ethereum (i.e. gas forwarding and data). 

> It differs in *two* key ways:

Reviewer 2: Probably merge this sentence and the one above. E.g.
"While Fuel does share some similar conceptual call paradigms to Ethereum (i.e. gas forwarding and data), it differs in *two* key ways:"

Author: Good call.

1) [**Native Assets**](./native_assets.md): FuelVM calls can forward any native asset not just Ether.

2) [**No Data Serialization**](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#vm-initialization): Fuel calls **do not** need to serialize data into ABI format, instead they simply pass pointers.

> This is because Fuel has a shared global memory context which all call frames can read from and so calling contracts only requires pointers to be passed, and no re-serialization of data is ncessarly.

Reviewer 2: ```suggestion
This is because Fuel has a shared global memory context which all call frames can read from and so calling contracts only requires pointers to be passed, and no re-serialization of data is necessarily.
```

> While Fuel does share some similar conceptual call paradigms to Ethereum (i.e. gas forwarding and data). 
> 
> It differs in *two* key ways:
> 
> 1) [**Native Assets**](./native_assets.md): FuelVM calls can forward any native asset not just Ether.
> 
> 2) [**No Data Serialization**](https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/main.md#vm-initialization): Fuel calls **do not** need to serialize data into ABI format, instead they simply pass pointers.
> 
> This is because Fuel has a shared global memory context which all call frames can read from and so calling contracts only requires pointers to be passed, and no re-serialization of data is ncessarly.

Reviewer: ```suggestion
While the Fuel contract calling paradigm is similar to Ethereum's (using an ABI, forwarding gas and data), it differs in two key ways:

1. [**Native assets**](./native_assets.md): FuelVM calls can forward any native asset not just base asset.
1. **No data serialization**: Contract calls in the FuelVM do not need to serialize data to pass it between contracts; instead they simply pass a pointer to the data. This is because the FuelVM has a shared global memory which all call frames can read from.
```
